### PR TITLE
GH action to check bundlesize of PR

### PR DESCRIPTION
**Pull request by katiegoines**
#### Description of changes:
GitHub action to check bundle size of below selected pages on PR branch vs on main branch. PR to fail if bundle size increases by +5% on any of the selected pages.

- `/cli`
- `/cli/start/install`
- `/lib/auth/getting-started/q/platform/[platform]`
- `/start`
- `/cli`

Example failure from testing on my local fork (with console logs):

Screenshot 2023-10-09 at 12 59 10 PM

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: .github/workflows/scripts/check_bundle_size.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      }
+    });
+    return headBundleSizes;
+  },
+
+  compareBundles: (baseBundles, headBundles) => {
+    for (let i = 0; i < baseBundles.length; i++) {
+      if (baseBundles[i] !== headBundles[i]) {
+        if (baseBundles[i].parsedSize * 1.05 < headBundles[i].parsedSize) {
+          return false;
```
Comment (jacoblogan) [inline .github/workflows/scripts/check_bundle_size.js:62]: should this be `return true` it looks like the final steps condition `if: ${{ steps.compare-bundles.outputs.result }}` is looking for a value of `true` to block the PR or is it just any non null value?

File: .github/workflows/scripts/check_bundle_size.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          totalStatSize: page.totalStatSize
+        });
+      }
+    });
+    return headBundleSizes;
+  },
+
+  compareBundles: (baseBundles, headBundles) => {
+    for (let i = 0; i < baseBundles.length; i++) {
+      if (baseBundles[i] !== headBundles[i]) {
```
Comment (jacoblogan) [inline .github/workflows/scripts/check_bundle_size.js:60]: Is the purpose of this `if` statement just to skip over the comparison if the same object is passed into the function as `baseBundles` and `headBundles`?  Are we guaranteed that the pages Array generated will always be in the same order or is there a situation where the filter function from the `checkBundleSize` function produces arrays that are in a different order between head and base (such as pages being added or removed)?

File: .github/workflows/scripts/check_bundle_size.js (hunk 1 above)
Comment (katiegoines) [reply .github/workflows/scripts/check_bundle_size.js:62]: It's looking for any non-null value, but `true` is better. Changing

File: package.json [hunk 3]
```diff
@@ -20,6 +20,7 @@
     "@emotion/react": "^11.1.5",
     "@emotion/styled": "^11.3.0",
     "@mdx-js/mdx": "^2.0.0",
+    "@next/bundle-analyzer": "^13.5.3",
```
Comment (timngyn) [inline package.json:23]: Could we move this into the `DevDependencies`? I think the workflow should still work if we move it there and that way this wouldn't be required for the "production" build

File: .github/workflows/scripts/check_bundle_size.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+module.exports = {
+  checkBaseBundleSize: () => {
+    const fs = require('fs');
+    const jsonString = fs.readFileSync('.github/analyze/bundles.json');
```
Comment (timngyn) [inline .github/workflows/scripts/check_bundle_size.js:4]: Sorry if I missed it somewhere, but is this file getting generated from a step or something like that?

Comment (timngyn) [reply .github/workflows/scripts/check_bundle_size.js:4]: Ah nvm, I see it now, you make the build script run with the `ANALYZE` env var, sorry about that!

File: .github/workflows/scripts/check_bundle_size.js (hunk 1 above)
Comment (katiegoines) [reply .github/workflows/scripts/check_bundle_size.js]: Fixed this logic to be able to console.log any problem pages.

File: .github/workflows/scripts/check_bundle_size.js (hunk 2 above)
Comment (katiegoines) [reply .github/workflows/scripts/check_bundle_size.js]: Fixed this logic to check more explicitly.

**Review by jacoblogan — APPROVED** [review]

**Review by timngyn — APPROVED** [review]

**Pull request merged**